lib/manager/components/CreateFeedSource.js

```diff
@@ -0,0 +1,234 @@
+// @flow
+
+import memoize from 'lodash.memoize'
+import React, {Component} from 'react'
+import update from 'react-addons-update'
+import {
+  Button,
+  Checkbox,
+  Col,
+  ControlLabel,
+  FormControl,
+  FormGroup,
+  Grid,
+  HelpBlock,
+  ListGroup,
+  ListGroupItem,
+  Panel,
+  Row
+} from 'react-bootstrap'
+import {isUri} from 'valid-url'
+
+import ManagerPage from '../../common/components/ManagerPage'
+import ProjectHeader from './ProjectHeader'
+import {validationState} from '../util'
+
+import type {Project, User} from '../../types'
+
+type NewFeed = {
+  autoFetchFeed: boolean,
+  deployable: boolean,
+  name: string,
+  projectId: string,
+  retrievalMethod?: string,
+  url: string
+}
+
+type Props = {
+  createFeedSource: NewFeed => void,
+  fetchProject: (string) => void,
+  project: Project,
+  projectId: string,
+  user: User
+}
+
+type Validation = {
+  _form: boolean,
+  name?: boolean,
+  url?: boolean
+}
+
+type State = {
+  model: NewFeed,
+  validation: Validation
+}
+
+export default class CreateFeedSource extends Component {
+  props: Props
+  state: State
+
+  componentWillMount () {
+    const {fetchProject, project, projectId} = this.props
+    if (!project) {
+      fetchProject(projectId)
+    } else {
+      this._initializeState(this.props)
+    }
+  }
+
+  componentWillReceiveProps (nextProps: Props) {
+    this._initializeState(nextProps)
+  }
+
+  _initializeState (props: Props) {
+    this.setState({
+      model: {
+        autoFetchFeed: false,
+        deployable: false,
+        name: '',
+        projectId: props.projectId,
+        url: ''
+      },
+      validation: {
+        _form: false,
+        name: true,
+        url: true
+      }
+    })
+  }
+
+  _onInputChange = memoize(
+    (fieldName: string) => (evt: Event & {target: HTMLInputElement}) => {
+      const updatedState: State = update(this.state, {
```

Comment:
Actually, I do think I want to use it here because I am modifying a subkey of a root key of the state.  In this case I'm modifying `this.state.model[fieldName]`.

Reply:
Ah, you're totally right.  I did not read that carefully.